flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/partition/JobAwareShuffleEnvironmentImpl.java

```diff
@@ -0,0 +1,194 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.runtime.taskexecutor.partition;
+
+import org.apache.flink.api.common.JobID;
+import org.apache.flink.metrics.MetricGroup;
+import org.apache.flink.runtime.deployment.InputGateDeploymentDescriptor;
+import org.apache.flink.runtime.deployment.ResultPartitionDeploymentDescriptor;
+import org.apache.flink.runtime.executiongraph.ExecutionAttemptID;
+import org.apache.flink.runtime.executiongraph.PartitionInfo;
+import org.apache.flink.runtime.io.network.api.writer.ResultPartitionWriter;
+import org.apache.flink.runtime.io.network.partition.PartitionProducerStateProvider;
+import org.apache.flink.runtime.io.network.partition.ResultPartitionID;
+import org.apache.flink.runtime.io.network.partition.consumer.InputGate;
+import org.apache.flink.runtime.shuffle.ShuffleEnvironment;
+import org.apache.flink.util.Preconditions;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.function.Consumer;
+import java.util.stream.Collectors;
+
+/**
+ * Wraps a {@link ShuffleEnvironment} to allow tracking of partitions per job.
+ */
+public class JobAwareShuffleEnvironmentImpl<G extends InputGate> implements JobAwareShuffleEnvironment<NotifyingResultPartitionWriter, G> {
+
+	private static final Consumer<ResultPartitionID> NO_OP_NOTIFIER = partitionId -> {};
+
+	private final ShuffleEnvironment<?, G> backingShuffleEnvironment;
+	private final PartitionTable inProgressPartitionTable = new PartitionTable();
+	private final PartitionTable finishedPartitionTable = new PartitionTable();
```

Comment:
Is it possible to not distinguish these two partition tables, only maintain one unreleased partition table?
That means as long as the partition is not released it would keep in this table, no matter this partition is in setup or finished state. Then we could simple check this table for occupying local resource.

Reply:
I wanted to ensure that the TM does not issue a release call for in-progress partitions; as these are already handled by the task cleanup when failing the partitions.

Reply:
see also the special cases section in the PR description.

Reply:
I got it, thanks for this explanation.
I think it is better to give some annotations here for this consideration of releasing finished partitions in order to make other guys clear while seeing the codes.